gcloud/_apitools/test_http_wrapper.py

```diff
@@ -0,0 +1,824 @@
+                          retries=3,
+                          max_retry_wait=WAIT,
+                          retry_func=_retry,
+                          wo_retry_func=_wo_exception,
+                          check_response_func=_checked.append)
+
+        self.assertEqual(len(_created), 3)
+        for attempt in _created:
+            self.assertEqual(attempt,
+                             ((HTTP, REQUEST), {
+                                'redirections': 5,
+                                'check_response_func': _checked.append,
+                             }))
+        self.assertEqual(_checked, [])  # not called by '_wo_exception'
+        self.assertEqual(len(_retried), 2)
+        for index, retry in enumerate(_retried):
+            self.assertTrue(retry.http is HTTP)
+            self.assertTrue(retry.http_request is REQUEST)
+            self.assertTrue(isinstance(retry.exc, _Retry))
+            self.assertEqual(retry.num_retries, index + 1)
+            self.assertEqual(retry.max_retry_wait, WAIT)
+
+
+class Test__RegisterHttpFactory(unittest2.TestCase):
+
+    def _callFUT(self, *args, **kw):
+        from gcloud._apitools.http_wrapper import _RegisterHttpFactory
+        return _RegisterHttpFactory(*args, **kw)
+
+    def test_it(self):
+        from gcloud._testing import _Monkey
+        from gcloud._apitools import http_wrapper as MUT
+        _factories = []
+
+        FACTORY = object()
+
+        with _Monkey(MUT, _HTTP_FACTORIES=_factories):
+            self._callFUT(FACTORY)
+            self.assertEqual(_factories, [FACTORY])
+
+
+class Test_GetHttp(unittest2.TestCase):
+
+    def _callFUT(self, *args, **kw):
+        from gcloud._apitools.http_wrapper import GetHttp
+        return GetHttp(*args, **kw)
+
+    def test_wo_registered_factories(self):
+        from httplib2 import Http
+        from gcloud._testing import _Monkey
+        from gcloud._apitools import http_wrapper as MUT
+        _factories = []
+
+        with _Monkey(MUT, _HTTP_FACTORIES=_factories):
+            http = self._callFUT()
+
+        self.assertTrue(isinstance(http, Http))
+
+    def test_w_registered_factories(self):
+        from gcloud._testing import _Monkey
+        from gcloud._apitools import http_wrapper as MUT
+
+        FOUND = object()
+
+        _misses = []
+
+        def _miss(**kw):
+            _misses.append(kw)
+            return None
+
+        _hits = []
+
+        def _hit(**kw):
+            _hits.append(kw)
+            return FOUND
+
+        _factories = [_miss, _hit]
+
+        with _Monkey(MUT, _HTTP_FACTORIES=_factories):
+            http = self._callFUT(foo='bar')
+
+        self.assertTrue(http is FOUND)
+        self.assertEqual(_misses, [{'foo': 'bar'}])
+        self.assertEqual(_hits, [{'foo': 'bar'}])
+
+
+class _Dummy(object):
+    def __init__(self, **kw):
+        self.__dict__.update(kw)
+
+
+class _Request(object):
+    __slots__ = ('url', 'http_method', 'body', 'headers', 'loggable_body',)
```

Comment:
I wouldn't be that defensive writing tests for new code, but `apitools` is more than a little whacky.